File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int solve(TreeNode root, int[] maxi){
        if(root == null)return 0;
        
        int left_sum  =  Math.max(0, solve(root.left, maxi));
        int right_sum = Math.max(0, solve(root.right, maxi));
        
        maxi[0] = Math.max(maxi[0], root.val + left_sum + right_sum);
        
        return Math.max(left_sum, right_sum) + root.val;
    }
    public int maxPathSum(TreeNode root) {
        int[] maxi = {Integer.MIN_VALUE};
        solve(root, maxi);
        return maxi[0];
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    private:
    int solve(TreeNode* root, int &mx){
        if(!root)return 0;
        int left = max(0, solve(root->left, mx));
        int right = max(0, solve(root->right, mx));
        mx = max(mx, left + right + root->val);
        return max(left, right) + root->val;
    }
public:
    int maxPathSum(TreeNode* root) {
        int mx = INT_MIN;
        
        return max(mx, solve(root, mx));
    }
};